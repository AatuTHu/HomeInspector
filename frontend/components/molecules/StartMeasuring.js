import { View, Text } from 'react-native'
import { styles } from "../../styles/styles"
import { useState, useEffect } from 'react'
import Buttons from '../atoms/Buttons'
import localVariables from '../../env'

export default function StartMeasuring({selected,setTempStarted,tempStarted,setHumStarted,humStarted}) {

  const { apiKey, temperatureStartURL, humidityStartURL } = localVariables
  const [ started, setStarted ] = useState(false);

  useEffect(() => {
    async function fetchIsStarted () {
      let url = selected === 'humidity' ? humidityStartURL : temperatureStartURL

      const response = await fetch(url).then((res)=>
       res.json()
     ).catch((error) => { 
       console.log(error.message); 
     })

     if(response !== undefined) {
        setStarted(response)
     }
    }
   fetchIsStarted()
  }, [])

  const startMeasuring = async() => {
    
      let mode = started ? 0 : 1
      let url = selected === 'humidity' ? humidityStartURL : temperatureStartURL

      const response = await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: `apiKey=${apiKey}&measuringMode=${mode}`,
      }).catch((error) => { 
        console.error(error.message); 
      }); 
  
      if(response.status === 200) {
        setStarted(!started)
        selected === 'humidity' ? setHumStarted(!humStarted) : setTempStarted(!tempStarted)
      } 
    }

  return (
    <View style = { styles.buttonRowContainer }>
     <Buttons text = "Start" textStyle="white" type = "start" onPress={ startMeasuring }/>
        <View style = { styles.statusBox }>
          <Text style = { styles.subHeadingText }>{ started ? (<Text style = {{"color" : "green"}}>On</Text>) : (<Text>Off</Text>)}</Text>
        </View>
    </View>
  )
}