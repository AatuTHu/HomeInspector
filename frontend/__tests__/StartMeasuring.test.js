import { render, fireEvent, waitFor } from '@testing-library/react-native';
import StartMeasuring from '../components/molecules/StartMeasuring';

global.fetch = jest.fn(() =>
Promise.resolve({
  status: 200,
  json: () => Promise.resolve({}),
})
);

describe('StartMeasuring', () => {

afterEach(() => {
    jest.clearAllMocks();
  })

it('renders correctly', () => {
    const screen = render(<StartMeasuring/>);
    expect(screen).toMatchSnapshot();
});

it('should start measuring temperature when response status is 200', async () => {

  const setTempStarted = jest.fn()
  const { getByText } = render(<StartMeasuring selected="temperature" setTempStarted={setTempStarted} tempStarted={false}/>);
  const startButton = getByText('Start');
  fireEvent.press(startButton);
  await waitFor(() => expect(setTempStarted).toHaveBeenCalledWith(true));
});

it('should start measuring humidity when response status is 200', async () => {

  const setHumStarted = jest.fn();
  const { getByText } = render(<StartMeasuring selected="humidity" setHumStarted={setHumStarted} humStarted={false} />);
  const startButton = getByText('Start');
  fireEvent.press(startButton);
  await waitFor(() => expect(setHumStarted).toHaveBeenCalledWith(true));
})
    
})