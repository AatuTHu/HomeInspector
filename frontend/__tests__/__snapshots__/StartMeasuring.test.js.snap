// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartMeasuring renders correctly 1`] = `
<View
  style={
    {
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 20,
      "marginTop": 1,
      "width": "95%",
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#03254E",
        "borderRadius": 20,
        "borderWidth": 1,
        "opacity": 1,
        "padding": 5,
        "width": "50%",
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      Start
    </Text>
  </View>
  <View
    style={
      {
        "backgroundColor": "#D9E5D6",
        "borderRadius": 22,
        "justifyContent": "center",
        "width": "30%",
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 22,
          "textAlign": "center",
        }
      }
    >
      <Text>
        Off
      </Text>
    </Text>
  </View>
</View>
`;
